Importing (some of) Standard.Base works from NI runner

JaroslavTulach opened the pull request:
### Pull Request Description

- part of `Standard.Base` is compiled into `runner` executable and can `IO.println` fast 
- proves #9774 is viable
- achieved by updating list of classes needing reflection by re-running the NI agent for `enso --run test/Base_Tests` 
- the updated configuration `.json` files are part of this PR
- including JAR files of `Standard.Base` library on `-cp` when building the `engine/runner/buildNativeImage`
- got `ENSO_JAVA=espresso ./runner` working again by avoiding `java.desktop` via `jlink`
- that requires changes to SnakeYaml - [send to upsteam](https://github.com/snakeyaml/snakeyaml/pull/12)
- to unblock this PR two Apache licensed SnakeYaml files are temporarily copied to our repository

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- [x] Integration test have been updated to use `Standard.Base` library

JaroslavTulach committed:
Adjust path to engine/runner

JaroslavTulach committed:
Regenerated with env JAVA_OPTS="-agentlib:native-image-agent=config-merge-dir=./engine/runner/src/main/resources/META-INF/native-image/org/enso/runner" enso --run test/Base_Tests/

JaroslavTulach committed:
Include all Standard/Base JARs on classpath when building NI of engine/runner

JaroslavTulach committed:
Importing from Standard.Base works in NI mode again

Akirathan commented on engine/runner/src/test/resources/Factorial.enso:
This file is run on native-image in CI, right? Via something like `./runner --run Factorial.enso`? I think it is time to increase the test coverage of native-image generated runner.

JaroslavTulach replied:
The coverage has increased significantly! We can so `+`, `*` and `IO.println` ;-)

Akirathan commented:
Are you going to experiment with other standard libraries, like `Standard.Table` as well in this PR? Or are you going to postpone other libs in different PRs? In any case, it would be nice to start increasing test coverage of native-image-produced `engine-runner`. We used to have only `Factorial.enso` that did not import anything from `Standard.Base`.

JaroslavTulach commented on build.sbt:
```diff
@@ -2351,7 +2351,9 @@ lazy val `engine-runner` = project
             "runtime.jar",
             "runner.jar"
-          ),
+          ) ++ (`base-polyglot-root`
```
Looks like this code is causing [sbt to fail to compile build script](https://github.com/enso-org/enso/actions/runs/8971799652/job/24641523363#step:5:20). It did compiled in incremental mode, but with clean repo it doesn't compile.

hubertp replied:
it might have worked on an already built project because jars were already built. This however does not specify the dependency between subprojects or at least not referencing the `package` target.

JaroslavTulach replied:
I see. Thank you 46b48d6da1 is the fix that prevents NPE.

JaroslavTulach replied:
With 2556a2c we first `buildEngineDistribution` and only then compute the `additionalCp`. I doubt _passing a thunk_ instead of a value is the right `sbt` way of doing this, but it seems to do the trick.

JaroslavTulach committed:
base-polyglot-root may not exist

JaroslavTulach commented:
> Are you going to experiment with other standard libraries, like `Standard.Table` as well in this PR? Or are you going to postpone other libs in different PRs? 

If this PR gets green, I'd like to integrated it as it [improves the usability a lot](https://github.com/enso-org/enso/pull/9866#discussion_r1591297640). However I don't claim this is a fix for #9774. The work will continue. I still want to create a `Feature` that will scan a library for `polyglot java import` and register such classes for reflection automatically.

> In any case, it would be nice to start increasing test coverage of native-image-produced `engine-runner`. 

Ideally, when we train the _native image agent_ on `test/Base_Tests` we should be able to run the tests on NI version. Right now something is broken. We want to fix it.

> We used to have only `Factorial.enso` that did not import anything from `Standard.Base`.

Now we have `Factorial.enso` that imports something from `Standard.Base`.

JaroslavTulach committed:
Compute additional path at invocation of the buildNativeImage task

JaroslavTulach committed:
Useful perf and debug switches

JaroslavTulach commented:
> If this PR gets green

All's well [including Factorial.enso invocation](https://github.com/enso-org/enso/actions/runs/8979950502/job/24662809301?pr=9866#step:7:3646) with [regular NI runner](https://github.com/enso-org/enso/actions/runs/8979950502/job/24662809301?pr=9866#step:7:1361), but...

there is a failure of [Factorial.enso invocation](https://github.com/enso-org/enso/actions/runs/8979950502/job/24662809301?pr=9866#step:7:3825) with [Espresso built NI runner](https://github.com/enso-org/enso/actions/runs/8979950502/job/24662809301?pr=9866#step:7:3773).

Looks like our Espresso support in NI is broken.

hubertp commented on build.sbt:
```diff
@@ -2348,10 +2348,18 @@ lazy val `engine-runner` = project
+              "runtime.jar",
+              "runner.jar"
+            )
+            val jars = `base-polyglot-root`.listFiles("*.jar")
```
That's really just delaying the problem but not solving it. It should be a proper dependency. I will take a look later today.

JaroslavTulach replied:
OK. However there is a proper [dependsOn](https://github.com/enso-org/enso/pull/9866/files#diff-5634c415cd8c8504fdb973a3ed092300b43c4b8fc1e184f7249eb29a55511f91R2382). E.g. at the end the system will behave correctly.

Akirathan replied:
You are probably right about the `dependsOn`. But still, I think there should be an assertion that all these JAR archives exists, and that the `base-polyglot-root` directory is not empty. A simple `assert` should be fine, in case of assertion error, sbt will display it.

JaroslavTulach replied:
The content of the `base-polyglot-root` directory now varies based on _regular vs. Espresso_ mode - we cannot assert exact content. Anyway we have CI tests that verify this all works in _both modes_.

hubertp reviewed (changes requested):
Rejecting (for now) because I don't want this PR to accidentally get in with the wrong sbt definition.

Akirathan commented on build.sbt:
```diff
@@ -2348,10 +2348,18 @@ lazy val `engine-runner` = project
-            "runtime.jar",
-            "runner.jar"
-          ),
+          additionalCp = () => {
```
I believe `additionalCp` does not need to be a function. You can just write something like:
```scala
additionalCp = {
  val core = Seq(...)
  val jars = Seq(...)
  core ++ jars
}
```
that will assign a proper sequence to `additionalCp`.

JaroslavTulach replied:
I don't think this works. I need to execute `buildEngineDistribution` first and only then search for the JAR files. Otherwise (with a clean build) they will not be present yet when the `{ ... }` code runs. with `() => { ... }` the code gets executed after the JARs are generated.

Akirathan commented on project/NativeImage.scala:
```diff
@@ -86,7 +86,7 @@ object NativeImage {
     initializeAtBuildtime: Seq[String]       = defaultBuildTimeInitClasses,
     mainClass: Option[String]                = None,
-    additionalCp: Seq[String]                = Seq(),
+    additionalCp: () => Seq[String]          = () => Seq(),
```
`additionalCp` does not need to be a function. See my [other comment](https://github.com/enso-org/enso/pull/9866/files#r1592000092). This is an unnecessary complication.

JaroslavTulach replied:
So far I am keeping as it is.

JaroslavTulach committed:
Merging removal of java.desktop to this branch

JaroslavTulach committed:
Configuration for NI debugging of the runner

JaroslavTulach committed:
NI is built from -cp - we need other method of detecting presence of Espresso

JaroslavTulach committed:
Script to generate 'small JDK' with native-image in jdk directory

JaroslavTulach committed:
Exclude Graal.js and Graal Python when using Espresso, as they don't run without java.desktop

JaroslavTulach committed:
Avoid special Graal.js settings

JaroslavTulach committed:
Patching SnakeYaml to run without java.desktop

JaroslavTulach commented on engine/runtime-instrument-common/src/main/java/org/yaml/snakeyaml/introspector/MethodProperty.java:
```diff
@@ -0,0 +1,185 @@
+ */
+package org.yaml.snakeyaml.introspector;
+
+// remove once https://github.com/snakeyaml/snakeyaml/pull/12 gets integrated
```
Remove this file once https://github.com/snakeyaml/snakeyaml/pull/12 gets integrated.

JaroslavTulach replied:
E.g. when upgrading to [SnakeYaml 2.3](https://bitbucket.org/snakeyaml/snakeyaml/wiki/Changes) to be released in ~August. Tracked as 
- #9113

JaroslavTulach committed:
Copy all JARs to polyglot/java directory in Espresso mode

JaroslavTulach committed:
Shelve SnakeYaml patches into a single project

Akirathan commented on engine/runner/src/main/java/org/enso/runner/ContextFactory.java:
```diff
@@ -149,7 +149,6 @@ PolyglotContext build() {
             .options(options)
             .option(RuntimeOptions.ENABLE_AUTO_PARALLELISM, Boolean.toString(enableAutoParallelism))
             .option(RuntimeOptions.WARNINGS_LIMIT, Integer.toString(warningsLimit))
-            .option("js.foreign-object-prototype", "true")
```
Why was this option here? Does this removal has anything to do with this PR or is this just a removal of something that should have been removed a long time ago already?

JaroslavTulach replied:
I don't know why this option was needed. As far as I know it doesn't influence any existing test. Some relict of the past.

Akirathan commented on project/GraalVM.scala:
```diff
@@ -51,27 +51,35 @@ object GraalVM {
-    "org.graalvm.shadowed" % "icu4j"              % version,
-    "org.graalvm.shadowed" % "xz"                 % version
-  )
+  val pythonPkgs = if ("espresso".equals(System.getenv("ENSO_JAVA"))) {
```
One of the mistakes I made when writing this code was that I have not specified this field as `private`. All of `pythonPkgs`, `jsPkgs`, and `espressoPkgs` can and *should* be private. I don't like the `if ("espresso"...)` condition repeated three times in this file. You can have this condition only on `langsPkgs` field that is actually public and is used from various projects in `build.sbt`.

JaroslavTulach replied:
Done in 2e3b0ed

Akirathan commented on project/NativeImage.scala:
```diff
@@ -86,22 +86,29 @@ object NativeImage {
+          s"$path\\bin\\native-image.cmd"
+        else s"$path/bin/native-image"
+
+      val localJdk = baseDirectory.value / ".." / ".." / "jdk"
```
This is supposed to pick up the minimal JDK downloaded with `jdk.sh`, right? Would be nic to embed the functionality of `jdk.sh` into sbt.

JaroslavTulach replied:
sbt uses the script to build the JDK since d29c9a1 - we build the `runner` executable only on Unix, so it may be sufficient to keep it this way for now.

JaroslavTulach committed:
Build smalljdk before buildNativeImage

JaroslavTulach committed:
Merge remote-tracking branch 'origin/develop' into wip/jtulach/NIwithBase9774

JaroslavTulach commented on engine/runner/src/test/resources/Factorial.enso:
```diff
@@ -1,13 +1,12 @@
-    fac number
+main number=5 =
+    v = fac number
+    IO.println v
```
Using `Standard.Base` library calls like `IO.println`

JaroslavTulach commented:
The important parts of this PR are green. Let's get ready for the integration.
- [Factorial.enso](https://github.com/enso-org/enso/pull/9866/files#r1601569448) is now using `Standard.Base` APIs
- `./built-distribution/enso-engine-*/enso-*/bin/enso --run engine/runner/src/test/resources/Factorial.enso` runs in **3225 ms**
- [./runner --run Factorial.enso](https://github.com/enso-org/enso/actions/runs/9094487679/job/24995689029?pr=9866#step:7:3631) uses classes compiled into native executable and runs in **1221 ms**
- [ENSO_JAVA=espresso ./runner --run Factorial.enso](https://github.com/enso-org/enso/actions/runs/9094487679/job/24995689029?pr=9866#step:7:3794) uses Espresso to execute Java code and it runs in **2859 ms**

We have _experimentally demonstrated_ compiling `Standard` libraries with `native-image` gives us **three times speedup** when starting and allows us to use `Standard.Base` from `./runner` executable to begin with. Let's integrate, unless there are objections!

JaroslavTulach committed:
Hide access to various Pkgs constants

hubertp commented on project/smalljdk.sh:
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash -x
```
This really doesn't belong here. It should be in `tools` and the right dependencies should be set between tasks.

hubertp reviewed (changes requested):
I don't like the changes in sbt. It very much doesn't feel like sbt'ish (not your fault) and adding some shell scripts into the project directory does not help things.  

Also, what I'm missing in the source code comments (or build documentation) is the concrete command on how to generate all native image config files. I know that can be found somehow via NI documentation but I'd like to reduce friction points.

hubertp committed:
Rewrite SBT support

- ditch bash script in favour of platform agnostic solution
- use proper dependencies between tasks to avoid dummy function calls

hubertp committed:
fix NPE

hubertp committed:
Correctly detect new javaHome for NI

hubertp committed:
nit

hubertp reviewed (approved):
I had a go at SBT and the changes are now in a more acceptable state. I know that the previous solution worked but the build setup easily starts to become a spaghetti monster when one starts to add random bash scripts.

@JaroslavTulach please add docs with an exact command on how to regenerate NI configs.

JaroslavTulach commented on build.sbt:
```diff
@@ -2350,6 +2352,55 @@ lazy val `engine-runner` = project
+          )
+      }
+
+      val exec =
```
Exec in Scala! Amazing, thank you @hubertp!

JaroslavTulach commented:
> I had a go at SBT and the changes are now in a more acceptable state. 

Thank you Hubert.

> @JaroslavTulach please add docs with an exact command on how to regenerate NI configs.

This commit contains the information 8cc1b2d

JaroslavTulach commented:
To regenerated Substrate VM reflection metadata as [this commit](8cc1b2d7e390b18e89176349eeb7d194025b406b) did, use:
```bash
JAVA_OPTS="-agentlib:native-image-agent=config-merge-dir=./engine/runner/src/main/resources/META-INF/native-image/org/enso/runner" enso --run test/Base_Tests/
```
